source/byte_buf.c

```diff
@@ -0,0 +1,109 @@
+/*
+* Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#include <aws/common/byte_buf.h>
+#include <assert.h>
+
+int aws_string_split_on_char(struct aws_byte_buf *input_str, char split_on, struct aws_array_list *output) {
+    assert(input_str);
+    assert(output);
+    assert(output->item_size >= sizeof(struct aws_byte_buf));
+
+    size_t current_pos = 0;
+    size_t last_offset = 0;
+
+    for(; current_pos < input_str->len; ++current_pos) {
+        if(current_pos < input_str->len - 2 && input_str->buffer[current_pos] == (uint8_t)split_on) {
+            struct aws_byte_buf buffer = {
+                    .buffer = input_str->buffer + last_offset,
+                    .len = current_pos - last_offset,
+            };
+
+            if (AWS_UNLIKELY(aws_array_list_push_back(output, (const void *)&buffer))) {
+                return AWS_OP_ERR;
+            }
+
+            last_offset = current_pos + 1;
+        }
+    }
+
+    if (last_offset < input_str->len) {
+        struct aws_byte_buf buffer = {
+                .buffer = input_str->buffer + last_offset,
+                .len = current_pos - last_offset,
+        };
+
+        if (AWS_UNLIKELY(aws_array_list_push_back(output, (const void *)&buffer))) {
+            return AWS_OP_ERR;
+        }
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+int aws_string_split_on_str(struct aws_byte_buf *input_str, char *split_on,
```

Comment:
What is the intended behavior when `input_str` contains NULs?

Reply:
If NULLS are in input str, we don't pay attention to them unless they're in the split criteria.

Reply:
The reason I ask is that `strstr` will not deal with embedded NULs.

Reply:
and I think that's a great point.... see below. I don't have a usecase in hand for supporting NULLs, and like the simplicity of applying YAGNI here.